### PR TITLE
Python API

### LINKED ISSUE #6197: Make CLI commands accessible as pure Python API
### Observed behavior

There seems to be some strange behavior for click where it shuts down the server automatically when there is not an active, user-created terminal running. This makes us unable to start and stop the Kolibri server from a GUI application such as the Mac app, and likely Android as well.

While there is a workaround we can use to sidestep click and call the Python functions directly,  this requires us to manually know how to initialize Kolibri and do so before calling each command, which may change over time and may eventually require different apps for different Kolibri versions.

This also means we cannot switch the apps to running a downloaded Kolibri pex file, as the problem occurs with the pex file as well. We need to extract the source and put it on the path to run it. 

It would be good to have a way to run the startup and shutdown process independently from click commands, either as pure python functions or an alternate cli mode. 

### Expected behavior

Starting Kolibri from within a GUI application by calling the user-facing startup commands works.

### User-facing consequences

Until this is resolved, users will be unable to update their Kolibri version without also updating the app.

### Errors and logs

…

### Steps to reproduce

Call `kolibri.utils.cli.main` or `kolibri.pex start --foreground` from inside a GUI application not running from the terminal.

### Context

Kolibri 0.13.0b3
macOS 10.14.5 (Mojave)
Embedded Safari (Not sure if version aligns with actual browser version)

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Consolidates non-CLI code in `kolibri.utils.main`
* Exposes core CLI methods as a Python API
* Adds a restart CLI function and API method

## References
Fixes #6197 

## Reviewer guidance
Test:
```
from kolibri.main import initialize
from kolibri.main import start
from kolibri.main import restart
from kolibri.main import enable_plugin
from kolibri.main import disable_plugin
```

Should we add anything else?
Is this the right place to expose these?

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/utils/cli.py [hunk 1]
```diff
@@ -450,13 +225,13 @@ def main(ctx):
 @main.command(cls=KolibriDjangoCommand, help="Start the Kolibri process")
 @click.option(
     "--port",
-    default=OPTIONS["Deployment"]["HTTP_PORT"],
+    default=None,
```
Comment (jredrejo) [inline kolibri/utils/cli.py:228]: Should we update https://kolibri.readthedocs.io/en/latest/access/index.html to comment the possibility of port autodiscovery using ---port=0 ?

File: kolibri/utils/main.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+    except (DatabaseError, SQLite3DatabaseError) as e:
+        if "malformed" in str(e):
+            logger.error(
+                "Your database appears to be corrupted. If you encounter this,"
+                "please immediately back up all files in the .kolibri folder that"
+                "end in .sqlite3, .sqlite3-shm, .sqlite3-wal, or .log and then"
+                "contact Learning Equality. Thank you!"
+            )
+        raise
```
Comment (jredrejo) [inline kolibri/utils/main.py:149]: I love this strategy :)

File: kolibri/utils/main.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                "Tried to check that the database was accessible "
+                "and an error occurred: {}".format(e)
+            )
+            raise
+
+
+def update(old_version, new_version):
+    """
+    Called whenever a version change in kolibri is detected
+    """
```
Comment (jredrejo) [inline kolibri/utils/main.py:237]: In the previous version of this function, the comment had also this text:

> TODO: We should look at version numbers of external plugins, too!

is there a reason to remove it now? I don't see any differences affecting the plugins

File: kolibri/utils/server.py [hunk 4]
```diff
@@ -580,13 +609,26 @@ def start(port=0, zip_port=0, serve_http=True, background=False):
         autoreloader = Autoreloader(bus)
         autoreloader.subscribe()
 
+    reload_plugin = RestartPlugin(bus)
```
Comment (jredrejo) [inline kolibri/utils/server.py:612]: I find `RestartPlugin` a confusing name. Reading plugins in the kolibri context does not bring magicbus to my mind at first

File: kolibri/utils/cli.py [hunk 5]
```diff
@@ -469,6 +244,10 @@ def start(port, zip_port, background):
     """
     Start the server on given port.
     """
+    port = OPTIONS["Deployment"]["HTTP_PORT"] if port is None else port
+    zip_port = (
+        OPTIONS["Deployment"]["ZIP_CONTENT_PORT"] if zip_port is None else zip_port
```
Comment (jredrejo) [inline kolibri/utils/cli.py:249]: As `OPTIONS["Deployment"]["ZIP_CONTENT_PORT"] `is 0 by default, the port number is not known if it's not set by the user. I think we should add this information to https://kolibri.readthedocs.io/en/latest/access/index.html so system administrators know they need to open this port in the firewall (thinking specially of the annoying messages coming from Windows on different ports everytime kolibri starts)

**Review by jredrejo — COMMENTED** [review]
Overall it looks good, most of the changes are just code relocated to a more logical place or where it makes kolibri start logic better. The few important changes are fine for my  pov and after testing it I have found no issues.
Just a few comments, specially related to documentation for our users and our code.

File: kolibri/utils/cli.py (hunk 1 above)
Comment (rtibbles) [reply kolibri/utils/cli.py:228]: I think that would be good, yes!

File: kolibri/utils/main.py (hunk 3 above)
Comment (rtibbles) [reply kolibri/utils/main.py:237]: No, except that it was addressed a long time ago, and I only just noticed that it is still here!

File: kolibri/utils/server.py (hunk 4 above)
Comment (rtibbles) [reply kolibri/utils/server.py:612]: Can update - but there are lots of classes called Plugin in this file!

File: kolibri/utils/cli.py (hunk 5 above)
Comment (rtibbles) [reply kolibri/utils/cli.py:249]: Yes, definitely - I'll add a user docs label to the PR.

File: kolibri/utils/server.py (hunk 4 above)
Comment (jredrejo) [reply kolibri/utils/server.py:612]: yes, and I feel the same in all the cases, but maybe it's too late to refactor all this stuff. They are actual plugins, even if they are not kolibri plugins. It's just made me think twice what I was checking becuase I'm quite biased on this field.

**Review by jredrejo — APPROVED** [review]

**Pull request merged**